File: daily-coding-problem/dcp-jvm/dcp-scala/src/main/scala/dcp/problems/_1_20/Problem11_AutoComplete.scala
package dcp.problems._1_20

import scala.annotation.tailrec
import scala.collection.mutable.ListBuffer

/**
 * Implement an autocomplete system. That is, given a query string s and a set of all possible query strings,
 * return all strings in the set that have s as a prefix.
 *
 * For example, given the query string de and the set of strings [dog, deer, deal], return [deer, deal].
 *
 * Hint: Try preprocessing the dictionary into a more efficient data structure to speed up queries.
 */

object Problem11_AutoComplete {

  class Trie() {
    val child: Array[Trie] = new Array[Trie](26)
    var word: Option[String] = None

    def append(word: String): Unit = {
      @scala.annotation.tailrec
      def add(curIdx: Int, trie: Trie): Unit = {
        if (curIdx == word.length) trie.word = Some(word)
        else {
          val char = word.charAt(curIdx) - 'a' // convert to int 0-25
          if (trie.child(char) == null) trie.child(char) = new Trie()
          add(curIdx + 1, trie.child(char))
        }
      }

      add(0, this)
    }

    def findByPrefix(prefix: String): Seq[String] = {
      val prefixNode = findPrefixNode(prefix, this)
      if (prefixNode != null) collectWordsAt(prefixNode).toSeq else Seq.empty[String]
    }

    @tailrec
    final def findPrefixNode(prefix: String, currTrie: Trie): Trie = {
      if (prefix.isEmpty) currTrie
      else {
        val idx = prefix.charAt(0) - 'a'
        if (currTrie == null || currTrie.child(idx) == null) null
        else findPrefixNode(prefix.substring(1), currTrie.child(idx))
      }
    }

    // dfs
    def collectWordsAt(trie: Trie, acc: ListBuffer[String] = ListBuffer[String]()): ListBuffer[String] = {
      if (trie.word.isDefined) acc.append(trie.word.get)
      else {
        for (i <- trie.child.indices) {
          if (trie.child(i) != null) collectWordsAt(trie.child(i), acc)
        }
        acc
      }
    }

  }

}

File: daily-coding-problem/dcp-py/1_20/problem11_auto_complete.py
"""
This problem was asked by Twitter.

Implement an autocomplete system. That is, given a query string s and a set of all possible query strings, return all
strings in the set that have s as a prefix.

For example, given the query string `de` and the set of strings [`dog`, `deer`, `deal`], return [`deer`, `deal`].

Hint: Try preprocessing the dictionary into a more efficient data structure to speed up queries.
"""


class Trie:

    def __init__(self):
        self.child = [None] * 26
        self.word = None

    def append(self, word):
        def add(curr_idx, trie):
            if curr_idx == len(word):
                trie.word = word
            else:
                char = ord(word[curr_idx]) - 97
                if not trie.child[char]:
                    trie.child[char] = Trie()
                add(curr_idx + 1, trie.child[char])

        add(0, self)

    def find_by_prefix(self, word):
        prefix_node = self.find_prefix_node(word, self)
        if prefix_node:
            return self.collect_words_at(prefix_node)

    def find_prefix_node(self, word, current_trie):
        if not word:
            return current_trie
        else:
            char = ord(word[0]) - 97
            return None if not current_trie or not current_trie.child[char] else \
                self.find_prefix_node(word[1:], current_trie.child[char])

    def collect_words_at(self, trie):
        def collect(current_trie, acc):
            if current_trie and current_trie.word:
                acc.append(current_trie.word)
            else:
                if current_trie and current_trie.child:
                    for i, trie_enum in enumerate(current_trie.child):
                        if current_trie.child[i]:
                            collect(current_trie.child[i], acc)
                return acc

        return collect(trie, [])


t = Trie()
words = ['dog', 'deer', 'deal']
[t.append(word) for word in words]

assert t.find_by_prefix('de') == ['deal', 'deer']
